src/number_theory/modular_forms/basic.lean

```diff
@@ -0,0 +1,286 @@
+end
+
+lemma slash_mul_SL2 (k1 k2 : ℤ) (A : SL(2,ℤ)) (f g : ℍ → ℂ) :
+  (f * g) ∣[k1 + k2, A] = (f ∣[k1, A]) * (g ∣[k2, A]) :=
+begin
+  have : (((↑ₘ(A : GL(2,ℝ)⁺)).det) : ℝ) = 1,
+  { simp only [coe_coe,matrix.special_linear_group.coe_GL_pos_coe_GL_coe_matrix,
+    matrix.special_linear_group.det_coe], },
+  have hs := slash_mul k1 k2 A f g,
+  simp_rw [this, one_smul] at hs,
+  convert hs,
+end
+
+lemma slash_mul_subgroup (k1 k2 : ℤ) (Γ : subgroup SL(2,ℤ)) (A : Γ) (f g : ℍ → ℂ) :
+  (f * g) ∣[k1 + k2, A] = (f ∣[k1, A]) * (g ∣[k2, A]) :=
+begin
+  have : (((↑ₘ(A : GL(2,ℝ)⁺)).det) : ℝ) = 1,
+  by { simp only [coe_coe,matrix.special_linear_group.coe_GL_pos_coe_GL_coe_matrix,
+    matrix.special_linear_group.det_coe], },
+  have t1 := slash_mul k1 k2 A f g,
+  rw this at t1,
+  simp only [coe_coe, one_smul] at t1,
+  convert t1,
+end
+
+/-- A function `f : ℍ → ℂ` is weakly modular, of weight `k ∈ ℤ` and level `Γ`, if for every matrix .
+ `γ ∈ Γ` we have `f(γ • z)= (c*z+d)^k f(z)` where `γ= ![![a, b], ![c, d]]`, and it acts on `ℍ`
+ via Moebius trainsformations. -/
+lemma weakly_modular_mem' (k : ℤ) (Γ : subgroup SL(2,ℤ)) (f : ℍ → ℂ) :
+  f ∈ (weakly_modular_submodule k Γ) ↔ ∀ γ : Γ, ∀ z : ℍ,
+  f (γ • z) = ((↑ₘγ 1 0 : ℝ) * z +(↑ₘγ 1 1 : ℝ))^k * f z :=
+begin
+  simp only [weakly_modular_mem],
+  split,
+  { intros h1 γ z,
+  have h3 : (f ∣[k, γ]) z = f z , by {simp_rw (h1 γ)},
+  rw [←h3, mul_comm],
+  simp only [slash_action_eq_slash, slash],
+  have h55 := zpow_neg_mul_zpow_self k (upper_half_plane.denom_ne_zero (γ : GL(2, ℝ)⁺) z),
+  simp only [upper_half_plane.denom, upper_half_plane.subgroup_to_sl_moeb, upper_half_plane.sl_moeb,
+    coe_coe, matrix.special_linear_group.coe_GL_pos_coe_GL_coe_matrix,
+    matrix.special_linear_group.coe_matrix_coe, int.coe_cast_ring_hom, matrix.map_apply,
+    of_real_int_cast] at *,
+  rw [mul_assoc, h55, ←int.coe_cast_ring_hom, ←matrix.special_linear_group.coe_matrix_coe,
+    matrix.special_linear_group.det_coe ((γ : SL(2, ℤ)) : SL(2, ℝ))],
+  simp only [of_real_one, one_zpow, mul_one], },
+  { intros hf γ,
+  simp_rw [slash_action_eq_slash],
+  ext1,
+  rw [slash, ←upper_half_plane.subgroup_moeb, (hf γ x), mul_comm],
+  have h55 := zpow_neg_mul_zpow_self k (upper_half_plane.denom_ne_zero (γ : GL(2, ℝ)⁺) x),
+  simp_rw upper_half_plane.denom at *,
+  simp only [matrix.special_linear_group.coe_GL_pos_coe_GL_coe_matrix, coe_coe,
+    matrix.special_linear_group.coe_matrix_coe, int.coe_cast_ring_hom, matrix.map_apply,
+    of_real_int_cast] at h55,
+  simp only [coe_coe, matrix.special_linear_group.coe_GL_pos_coe_GL_coe_matrix,
+    matrix.map_apply, of_real_int_cast],
+  rw (matrix.special_linear_group.det_coe ((γ : SL(2, ℤ)) : SL(2, ℝ))),
+  simp only [matrix.special_linear_group.coe_matrix_coe, int.coe_cast_ring_hom, matrix.map_apply,
+    of_real_int_cast, of_real_one, one_zpow, mul_one],
+  simp_rw [← mul_assoc, h55],
+  simp },
+end
+
+lemma mul_modular (k_1 k_2 : ℤ) (Γ : subgroup SL(2,ℤ)) (f g : ℍ → ℂ)
```

Comment:
The many underscores which occur later when using this are a hint that `k_1`, `k_2`, `Γ`, `f` and `g` should be implicit.